### ISSUE TITLE
Big Bug

### ISSUE THREAD

**Issue by DebrisHauler** [AEFeinstein/Super-2024-Swadge-FW#221]
I've made good progress on the foundation. I'm adding some notes about the design. It's all more more locked down at this point as it definitely takes after **Dig Dug** primarily (not breakout).

The game will draw from a few mechanics I enjoyed in _Steamworld Dig_, _Dome Keeper_, _Helldivers_, and _Dave the Diver_. 

**Art style:**

**_I will take all the help I can get!_** 16-bit feel with a comically angry/evil playable Dr. Garbotnik character. I could especially use help with some backgrounds, a few different poses & sizes of garbotnik (he will have a few monologues in between gameplay sections), small expendable items ("wiles"), and... bugs. Please hit me up on Slack if you'd tackle any one of those or additional art that fits into the design.

**The gameplay loop:**
1. In the death dumpster, set your loadout by assigning "wiles" into the A, B, and C buttons. I am declaring the Death Dumpster our Magfest version of the Death Egg. Also think of it like the mothership in helldivers.
2. Go diving into the city dump to kill ants, beetles, roly-polys, etc... Return to the surface before time runs out, or before health runs out (to be determined...)?
3. Return to the death dumpster and prepare a small menu of recipes for the evening.
4. Set the death dumpster on Earth disguised as the "Big Bug" fast food joint. I need a little animation of a tarp being unfurled over the facade of the dumpster painted like a restaurant.
5. Serve pangolin customers bug meals to earn rings (similar minigame and camera perspective as in Dave the Diver). Fyi, pangolins do eat ants IRL. Pangolin customers look fairly realistic, crawling into the restaurant on all fours before sitting at the bar like a human. There is a very very low chance that a customer is the obviously cartoony style pango the pangolin walking in on two legs. In this case, you can launch the death dumpster back into space to get the end of game cutscene. Otherwise, customers clear out before night, and you launch into space alone.
6. Purchase more "wiles" with your rings which help the player to dive deeper and kill more bugs.

**Dr. Ovo Garbotnik's Wiles:**

Below are listed the in-game description text, plus my own notes. The player may need to try them out in various situations to learn how best to use them. And there should be some fun in discovering which wiles can synergize together. Wiles also do fancy things to the swadge LEDs when the player has the privilege of using one. With a few exceptions, they all work by dropping a small beacon device that marks where the thing will drop in from the sky. The player may call it down a vertically carved shaft, or go through some horizontal shaft and try targeting some bugs that are above the player.

_Faulty Wile_
-In-game text: Tends to explode in a few short seconds before establishing communication with the Death Dumpster. Still ironing out the kinks.
-Developer notes: Places it right at Garbotnik like all the other wiles, but this one just explodes right there. If you don't leave the blast radius before detonation, you will be pushed by the blast and take damage (unless shield is active). The blast also kills bugs and destroys dirt in the radius.

_Drill Bot_
-In-game text: A robot that drills horizontally. Very obedient in this task until the battery runs out.
-Developer notes: This is basically drillbert from _Domekeeper_. But it breaks down after a time in keeping with the overall design of wiles helping out for a brief period.

_Orbital Drill_
-In-game text: The Death Dumpster's laser will drill vertically, vaporizing targets below.
-Developer notes: Be sure to move horizontally out of the way before it gets to you. This can be useful to quickly carve deeper into the more dense sections, or call it down on one of the beefier bugs, or on several bugs vertically aligned in different layers.

_Shield_
-In-game text: Quantum particles are delivered (even through dirt) until a wearer is found who becomes invulnerable for 10 seconds.
-Developer notes: It traverses vertically down through any dirt. The player needs to intercept it to get the shield. If a bug intercepts it, then woops - the bug now has a shield. Useful not just for bumping bugs, but also bumping dirt without taking damage. Protects against the player's own damage-heavy wiles, or just helps offer peace of mind to safely return to the surface.

_501Kg Bomb_
-In-game text: Calls down ordinance from the Death Dumpster's stockpile.
-Developer notes: Very similar to the faulty wile with a larger blast radius. But this one drops down from space, so the player might want to call through a vertically carved shaft to detonate deep down.

_Atmospheric Atomizer_
-In-game text: It was supposed to be the Death Dumpster's own onboard time machine, but it does eviscerate the atmosphere within a few square miles for 30 seconds.
-Developer notes: Basically it turns off drag in the game, allowing you glide around freely bumping into lots of dirt tiles. Maybe especially interesting by launching yourself with a bomb while you have a shield applied. I can imagine some fun, sparkly swadge LED effects during this one.

_Doug's Improved Harpoon_
-In-game text: After finding this tool trashed by a resigned lunch-bucket worker, Dr. Ovo made modifications so it will more effectively neutralize any bug.
-Developer notes: This is useful to kill bugs without having to bump them and damage yourself. And a classic nod to Dig Dug.

https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/c453cf5d-4b05-4275-808e-d37aff1f4a18

**Comment by DebrisHauler** [comment]
https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/a480e639-f38a-4b2b-a682-0f4cdc8c33d9

**Comment by DebrisHauler** [comment]
https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/85d123fa-2a40-4591-ba9f-6d1667f82686

**Comment by DebrisHauler** [comment]
I need an artist's help for a tileable cave background png that is 255x255 resolution.

**Comment by DebrisHauler** [comment]
Design change to the control scheme: why not just go full helldivers and hold down the 'a' button to type in a stratagem sequence on the d-pad. That leaves 'b' button available for shooting harpoons, and 'c' button for tethering/untethering bugs.

**Comment by DebrisHauler** [comment]
I did a rebase which pulled in the 2025 sonic themed menu! It also fixed the row of dirt tiles at the top of the screen.
https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/b444451f-5d18-49e1-89ff-cf307496f7dc

**Comment by AEFeinstein** [comment]
I'm not done working with the visuals of the menu, but it's not gonna change _too_ drastically from here.

**Comment by DebrisHauler** [comment]
Did some nice tile work today in aseprite. 
video:
https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/c6f20cb7-984c-4d14-aedf-3ff4be17ecf4

Here is also an image of where the medium and hard densities are headed. They should all fit side by side nicely... eventually.

images:
![garbageDemo](https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/a873d2fe-83c5-4318-a9cf-920905893994)
![garbageDemoGrid](https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/1e3d72a4-d2be-44dd-89cd-fb775b7f0523)

**Comment by DebrisHauler** [comment]
> I'm not done working with the visuals of the menu, but it's not gonna change _too_ drastically from here.

I appreciate your involvement! It would be cool if the balls orbit around the center of the main menu!

**Comment by AEFeinstein** [comment]
Yep, working on animations, better fonts, better colors, and a better battery indicator

**Comment by DebrisHauler** [comment]
Progress with some shading and the medium hardness tileset.
video:
https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/aedb3731-0f99-4419-8276-5717c041e55f

**Comment by AEFeinstein** [comment]
That's looking so trashy!

**Comment by DebrisHauler** [comment]
I made a procedural way to generate levels in blender. The pixels are read in by the game to set soft, medium, or hard density tiles.

![image](https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/c70695a8-6adf-454c-9d35-3041de79ffc5)

While it's great to see each tile set works seamlessly beside the others, they are as visually distinctive as I was hoping. For gameplay purposes it would have been good if the player could very clearly distinguish blocks that will require more damage to delete. hmmmmmmmmm

https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/4362add5-4045-4612-afac-4e2e079695e6

**Comment by AEFeinstein** [comment]
Color could be a strong indicator, and the tiles seem pretty grey. There's some nice goop already which could be jazzed up.

**Comment by DebrisHauler** [comment]
> Color could be a strong indicator, and the tiles seem pretty grey. There's some nice goop already which could be jazzed up.

I might do something more with the goop. Using the same color palette is kind of important for the illusion of objects crossing the tile seams.

**Comment by DebrisHauler** [comment]
Did a despeckle filter on all the tiles. Got Ovo bouncing nicely. Digging only occurs at steep angles. Camera is no longer nauseating while you dig.

https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/940fed20-077b-4bb1-9862-25609e6e2092

**Comment by AEFeinstein** [comment]
For color, as long as the seams are consistently grey, color inside the tile can be whatever. Just a thought. Or you could make a ton of tiles for each transition at each edge. That seems less tenable.

**Comment by AEFeinstein** [comment]
Maybe have the tiles change in density as you chip away at them too? Might be nice feedback

**Comment by DebrisHauler** [comment]
> For color, as long as the seams are consistently grey, color inside the tile can be whatever. Just a thought. Or you could make a ton of tiles for each transition at each edge. That seems less tenable.

Good idea, I'll try eventually. Thank you! I may move on to other gameplay pillars first next jam session.

**Comment by DebrisHauler** [comment]
https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/8057f9c3-2eb5-4dcf-a706-9a044db109d5

**Comment by DebrisHauler** [comment]
Did lots of work on tile art today shown in the video above. The midground tiles will be regions restraining x,y traversing bugs. If you expose an entire midground region from the foreground, I plan a flood fill style collapse neutralizing any bugs there.

I've also started some animation in blender for when you bust tiles. Check it out!

https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/91450582-2e41-453e-ae89-b95c996b4c92

**Comment by AEFeinstein** [comment]
Whoa animation. I'm all programming, no art, so this is very impressive to me!

**Comment by DebrisHauler** [comment]
Boom, animated entities!

https://github.com/AEFeinstein/Super-2024-Swadge-FW/assets/6497989/e77fdbcb-8e25-4f8c-8e9e-dd43849a93dd

**Comment by DebrisHauler** [comment]
I'm trying to think up a method to crumble midground regions when they become unsupported by foreground. An exhaustive floodfill upon every midground reveal sounds expensive. Thinking of some behaviors that are present in valheim. Either managing the search over many frames or storing some support data in every tile. I don't know yet.

**Comment by DebrisHauler** [comment]
Terrain crumbling is a fairly important mechanic to me because it would open up different strategies to deal with enemies.

**Comment by AEFeinstein** [comment]
Hm, I don't quite get what you mean by midground and foreground. Is the game taking place on one 2D plane or more?

**Comment by DebrisHauler** [comment]
There is a another plane of gameplay rendered behind the player. Right now it's looking like a pale gray. So I was thinking of some bugs that can traverse x & y axes on that plane using a top down rendering. And some bugs that can only patrol left and right on the foreground using a side rendering.

**Comment by AEFeinstein** [comment]
If I'm understanding, you have two 2D grids that are somehow connected along the Z axis, and you're looking for wholly disconnected sections to crumble? Flood fill is going to be your best bet, and I would strongly recommend implementing it with a `list_t` rather than recursion. Recursion has a lot of function calling overhead (slow) and can chew through the stack (big). With a `list_t` you can:
* Add your initial index to the `list_t`
* While the `list_t` is not empty:
    * Remove an index from the `list_t`
    * Check that index, and mark it as checked & connected (every space needs a boolean flag)
    * Add all that index's unchecked neighbors to the `list_t`

**Comment by DebrisHauler** [comment]
Sounds spot on. Will be a while again for me to work on things. My main drive has died today and I'll have to setup everything again from scratch.

**Comment by AEFeinstein** [comment]
Oh yikes. I hope you had everything pushed to GitHub

**Comment by DebrisHauler** [comment]
Well this project is in a decent place. I only lost some blender and aseprite work for tilesets and level gen since I was only checking in final assets.

**Comment by DebrisHauler** [comment]
I'm overdue for a blog post! TLDR: I got my dev environment set up again, and even got to debugging on physical swadge hardware over UART & PuTTy thanks to all the other super helpful advice I'm picking up from Adam, Dylan, CNLohr, and others too. That is not easy. Got some awesome graphics & lighting done on the background, midground, and foreground. And got character rotation implemented.

There is some awesome new background art. Concept art was provided by Kaitie, and I went over it in Krita with a two tone lighting style, and dithered gradient in the lower depths with Python Image Library. Also came up with another background, so there are two layers back there at different parallax distances if you look closely.

This cycle was a big push towards immersive lighting, and I am excited to show you! I will be going forward with a battery stat the player must manage which doubles as health (rather than having a health bar). You will only dive as long as your headlamp is powered on to show you the way and your battery also keeps your machine airborne. Taking damage would also knock down your time remaining in the level by "routing energy into the shield".

I made this lookup texture where the red, green, and blue channels represent how the headlamp illuminates the level as garbotnik rotates from left to right. It represents left facing light, forward facing, and right facing. The left and right facing are also positioned to light up a little bit behind the player. Maybe this could be attributed to bounce light, but mainly I thought it would be frustrating gameplay to have complete darkness directly to your left or right. Technically I might have sampled the red channel backwards for right-facing situations, but this was a bit quicker to code using the three channels. As you rotate towards center, the midground interpolates more or less interpolates towards darkness, and then the foreground interpolates towards the blue channel and offsets that circular lookup left or right based on character rotation.

https://github.com/user-attachments/assets/571a50f1-a1c0-49b2-9b54-86508d4ee8ab

![headlampLookup](https://github.com/user-attachments/assets/156380fc-d659-43ec-8313-19ea51095119)

You may notice the tiles are lighting on a quarter-tile size with a dithering effect. It was a real puzzle to figure out this effect with the limited swadge color space, but I used python image library to chop up my full-size tiles and generate 6 carefully dithered versions of everything for brightness levels zero through five. This produced 1,080 files totaling 203 KB even after optimizing the tile lookups to reuse certain sprites. (We get 4 or 16 MB of space, right? This should be OK). The code loads them into spiram via for loops (which I am told I basically cannot possibly fill via game dev). This takes a good minute on the swadge, or 6 seconds in the emulator. I'll ask for some help adding a loading screen most likely.

![image](https://github.com/user-attachments/assets/3e505b11-993d-4b53-8327-2262e54bebb1)

Here's what the midground and background look like!

https://github.com/user-attachments/assets/a32f99f5-f4d3-4d25-b736-425a914844e3

The garbage crumbling algorithm or twin-stick shooting may be next on my list. Crumbling is halfway done already, but it is still needing a* pathfinding to determine stability and some management of uncalculated blocks. Twin stick shooting I imagine will consist of harpoons launching from one of those springy accordion comic hands that extends from behind the player sprite in the direction of a touchpad release event.

Here is a sneak peak of some bugs and pangolin customer animation from Kaitie. I'd like to run them through PIL to generate the different lighting levels if space & load time will allow. I might replace some pixels with bright colors because these would be the fully-lit versions, with 5 darker versions also generated.

![bu01 (1)](https://github.com/user-attachments/assets/84623466-024b-46c0-add8-58ca141431a7)
![bu02 (1)](https://github.com/user-attachments/assets/d19f2782-636e-493b-a323-524d22da4cf6)
![bu03 (1)](https://github.com/user-attachments/assets/54563433-99ae-4977-8688-92b7778dfae6)
![bu04 (1)](https://github.com/user-attachments/assets/593e8934-f67d-46df-8b9f-dd6e508f9354)
![bu05 (1)](https://github.com/user-attachments/assets/32df8856-e1e5-41b1-aabd-93e2554f45ed)
![pa00 (1)](https://github.com/user-attachments/assets/9ff6fa2d-a80a-49a3-a6b7-6800c0a23ea7)
![bu08 (1)](https://github.com/user-attachments/assets/f86f6fdb-ba75-46b8-ba32-6827a3e1866b)
![bu09 (1)](https://github.com/user-attachments/assets/f42efa92-4dc9-4508-bf37-8da1101de127)
![bu07 (1)](https://github.com/user-attachments/assets/a800cada-c4fe-42ad-8aba-339e4cb356e6)
![bu06 (1)](https://github.com/user-attachments/assets/278e938b-cafd-48a6-a19b-8b88808bbec5)

I look forward to hosting a midi music jam in October once the game is in a further state. And development may continue to be in weird concentrated bursts as I also handle some IRL things.

**Comment by AEFeinstein** [comment]
I've never been so excited to have this many bugs in the firmware.

**Comment by AEFeinstein** [comment]
Gonna keep this open until at least `bb_EnterMode()` is used instead of `bb_EnterModeSkipIntro()`
